File: django_server/final/app/schema.py
import graphene
import graphene_django

from django.contrib.auth.backends import UserModel
from django.db.models import Avg

from .models import Book, HasRead, rate_book, delete_rating


class UserType(graphene_django.DjangoObjectType):
    is_admin = graphene.Boolean()
    average_rating = graphene.Float()

    def resolve_is_admin(self, info):
        return self.is_staff

    def resolve_average_rating(self, info):
        query = self.books_read.all().aggregate(Avg('rating'))
        return query['rating__avg']

    class Meta(object):
        model = UserModel
        only_fields = ('id', 'username', 'books_read')


class BookType(graphene_django.DjangoObjectType):
    class Meta(object):
        model = Book


class HasReadType(graphene_django.DjangoObjectType):
    class Meta(object):
        model = HasRead


pagination_args = {
    "first": graphene.Int(default_value=10), 
    "offset": graphene.Int()
}


class Query(graphene.ObjectType):
    users = graphene.List(UserType, **pagination_args)
    user = graphene.Field(UserType, id=graphene.ID(required=True))
    books = graphene.List(BookType, fiction=graphene.Boolean(), **pagination_args)

    def resolve_users(self, info, **kwargs):
        first = kwargs.get('first')
        offset = kwargs.get('offset', 0)

        q = UserModel.objects.all()

        return q[offset:offset+first]

    def resolve_user(self, info, **kwargs):
        user_id =kwargs['id']
        return UserModel.objects.get(id=user_id)

    def resolve_books(self, info, **kwargs):
        fiction = kwargs.get('fiction')
        first = kwargs.get('first')
        offset = kwargs.get('offset', 0)
        q = Book.objects.all()

        if fiction is not None:
            q = q.filter(fiction=fiction)

        return q[offset:offset+first]


class RateBook(graphene.Mutation):
    class Arguments(object):
        book = graphene.Int(required=True)
        user = graphene.Int(required=True)
        rating = graphene.Int(required=True)

    has_read = graphene.Field(HasReadType)

    def mutate(self, info, **kwargs):
        book_id = kwargs['book']
        user_id = kwargs['user']
        rating = kwargs['rating']

        has_read = rate_book(book_id, user_id, rating)
        return RateBook(has_read=has_read)


class DeleteRating(graphene.Mutation):
    class Arguments(object):
        book = graphene.Int(required=True)
        user = graphene.Int(required=True)

    query = graphene.Field(Query)

    def mutate(self, info, **kwargs):
        book_id = kwargs['book']
        user_id = kwargs['user']

        delete_rating(book_id, user_id)
        return DeleteRating(query=Query)


class Mutation(graphene.ObjectType):
    rate_book = RateBook.Field()
    delete_rating = DeleteRating.Field()


schema = graphene.Schema(query=Query, mutation=Mutation)

File: node_server/final/src/schema.js
const graphql = require('graphql');
const knex = require('../db');
const {rateBook, deleteRating} = require('./models');

const UserType = new graphql.GraphQLObjectType({
    name: 'User',
    description: 'A user in the system',
    fields: () => {
        return {
            id: {
                type: graphql.GraphQLID,
                resolve(user) {
                    return user.id;
                }
            },
            username: {
                type: graphql.GraphQLString,
                resolve(user) {
                    return user.username;
                }
            },
            isAdmin: {
                type: graphql.GraphQLBoolean,
                resolve(user) {
                    return user.role === 'admin';
                }
            },
            booksRead: {
                type: graphql.GraphQLList(HasReadType),
                resolve(user) {
                    return knex('hasRead').where('userId', user.id);
                }
            },
            averageRating: {
                type: graphql.GraphQLFloat,
                async resolve(user) {
                    let query = await knex('hasRead')
                    .where('userId', user.id)
                    .avg('rating as avg_rating')
                    .first();

                    return query['avg_rating'];
                }
            }
        }
    }
});

const BookType = new graphql.GraphQLObjectType({
    name: 'Book',
    fields: () => {
        return {
            id: {
                type: graphql.GraphQLID,
                resolve(book) {
                    return book.id;
                }
            },
            title: {
                type: graphql.GraphQLString,
                resolve(book) {
                    return book.title;
                }
            },
            author: {
                type: graphql.GraphQLString,
                resolve(book) {
                    return book.author;
                }
            },
            publishedYear: {
                type: graphql.GraphQLInt,
                resolve(book) {
                    return book.publishedYear;
                }
            },
            fiction: {
                type: graphql.GraphQLBoolean,
                resolve(book) {
                    return book.fiction;
                }
            },
            readBy: {
                type: graphql.GraphQLList(HasReadType),
                resolve(book) {
                    return knex('hasRead').where('bookId', book.id);
                }
            }
        }
    }
});

const HasReadType = new graphql.GraphQLObjectType({
    name: 'HasRead',
    fields: {
        book: {
            type: BookType,
            resolve(hasRead) {
                return knex('book').where('id', hasRead.bookId).first();
            }
        },
        rating: {
            type: graphql.GraphQLInt,
            resolve(hasRead) {
                return hasRead.rating;
            }
        },
        user: {
            type: UserType,
            resolve(hasRead) {
                return knex('user').where('id', hasRead.userId).first();
            }
        }
    }
});

const paginationArgs = {
    first: {
        type: graphql.GraphQLInt,
        defaultValue: 10
    },
    offset: {
        type: graphql.GraphQLInt
    }
};

const QueryType = new graphql.GraphQLObjectType({
    name: 'Query',
    fields: {
        users: {
            type: graphql.GraphQLList(UserType),
            args: paginationArgs,
            resolve(root, args) {
                let query = knex('user');

                const limit = args.first;
                const offset = args.offset;

                if (limit) {
                    query = query.limit(limit);
                }

                if (offset) {
                    query = query.offset(offset);
                }

                return query;
            }
        },
        user: {
            type: UserType,
            args: {
                id: {
                    type: graphql.GraphQLNonNull(graphql.GraphQLID)
                }
            },
            resolve(root, args) {
                return knex('user').where('id', args.id).first();
            }
        },
        books: {
            type: graphql.GraphQLList(BookType),
            args: {
                fiction: {
                    type: graphql.GraphQLBoolean
                }, ...paginationArgs
            },
            resolve(root, args) {
                let query = knex('book');
                console.log(args);
                if (args.fiction != null) {
                    query = query.where('fiction', args.fiction);
                }

                const limit = args.first;
                const offset = args.offset;

                if (limit) {
                    query = query.limit(limit);
                }

                if (offset) {
                    query = query.offset(offset);
                }
                return query;
            }
        }
    }
});

const MutationType = new graphql.GraphQLObjectType({
    name: 'Mutation',
    fields: () => {
        return {
            rateBook: {
                type: HasReadType,
                description: 'Rate a book or update a rating',
                args: {
                    user: {
                        type: graphql.GraphQLNonNull(graphql.GraphQLID)
                    },
                    book: {
                        type: graphql.GraphQLNonNull(graphql.GraphQLID)
                    },
                    rating: {
                        type: graphql.GraphQLID
                    },
                },
                resolve(source, args) {
                    const userId = args.user;
                    const bookId = args.book;
                    const rating = args.rating;

                    return rateBook(bookId, userId, rating);
                }
            },
            deleteRating: {
                type: QueryType,
                description: 'Delete a book rating',
                args: {
                    user: {
                        type: graphql.GraphQLNonNull(graphql.GraphQLID)
                    },
                    book: {
                        type: graphql.GraphQLNonNull(graphql.GraphQLID)
                    }
                },
                resolve(source, args) {
                    const userId = args.user;
                    const bookId = args.book;

                    deleteRating(bookId, userId);
                    return QueryType;
                }
            }
        }
    }
});

const schema = new graphql.GraphQLSchema({query: QueryType, mutation: MutationType});
module.exports = schema;